test/dialects/dialects_test.py

```diff
@@ -22,57 +25,123 @@
     # Load the right dialect
-    config = FluffConfig(overrides=dict(dialect=dialect))
+    config = FluffConfig(overrides=json.loads(config_overrides))
     tokens, lex_vs = Lexer(config=config).lex(raw)
     # From just the initial parse, check we're all there
     assert "".join(token.raw for token in tokens) == raw
     # Check we don't have lexing issues
     assert not lex_vs
-
     # Do the parse WITHOUT lots of logging
     # The logs get too long here to be useful. We should use
     # specific segment tests if we want to debug logs.
-    if raw:
-        parsed = Parser(config=config).parse(tokens)
-        print(f"Post-parse structure: {parsed.to_tuple(show_raw=True)}")
-        print(f"Post-parse structure: {parsed.stringify()}")
-        # Check we're all there.
-        assert parsed.raw == raw
-        # Check that there's nothing unparsable
-        typs = parsed.type_set()
-        assert "unparsable" not in typs
+    if not raw:
+        return None
+
+    return Parser(config=config).parse(tokens)
+
+
+@pytest.fixture(scope="session")
+def cached_parser():
+    """Creates a cached version of the Lex and Parse.
+
+    This drastically reduces execution time while allowing nicbe test naming.
+    """
+    # Manually apply decorator to allow the use of _lex_and_parse
+    # ouside this fixture.
+    cached_parse_fn = lru_cache(maxsize=None)(_lex_and_parse)
+
+    def lex_and_parse(
+        config_overrides: Dict[str, Any], raw: str
+    ) -> Optional[BaseSegment]:
+        """Sets up cache parsable inputs from normal inpurts."""
+        cachable_str = json.dumps(config_overrides, sort_keys=True)
+        return cached_parse_fn(cachable_str, raw)
+
+    return lex_and_parse
+
+
+@pytest.mark.parametrize("dialect,file", parse_success_examples)
+def test__dialect__base_file_parse(dialect, file, cached_parser):
+    """For given test examples, check successful parsing."""
+    raw = load_file(dialect, file)
+    config_overides = dict(dialect=dialect)
+    # Use the helper function to avoid parsing twice
+    parsed: Optional[BaseSegment] = cached_parser(config_overides, raw)
+    if not parsed:
+        return
+
+    print(f"Post-parse structure: {parsed.to_tuple(show_raw=True)}")
+    print(f"Post-parse structure: {parsed.stringify()}")
+    # Check we're all there.
+    assert parsed.raw == raw
+    # Check that there's nothing unparsable
+    typs = parsed.type_set()
+    assert "unparsable" not in typs
+
+
+@pytest.mark.integration_test
+@pytest.mark.parametrize("dialect,file", parse_success_examples)
+def test__dialect__base_broad_fix(
+    dialect, file, raise_critical_errors_after_fix, cached_parser
+):
+    """Run a full fix with all rules, in search of critical errors."""
+    raw = load_file(dialect, file)
+    config_overides = dict(dialect=dialect)
+    # Lean on the cached result of the above test if possible
+    parsed: Optional[BaseSegment] = cached_parser(config_overides, raw)
```

Comment:
Will this every be cached now they are run separately? Wondering if it's costing us time to cache these (only to use them once), than it would be not to have the cache?

Reply:
Yup I'm look if its worth having. I had a few thoughts
- isn't helping across runs (obviously)
- still help for local testing
- I noticed another place where we are parsing the same files again but on a 2sec fix it didnt work

Reply:
I removed it for now